src/server_manager/web_app/cloud_accounts.ts

```diff
@@ -12,28 +12,130 @@
-    return this.digitalOceanAccountFactory(accessToken);
+    this.digitalOceanAccount = this.createDigitalOceanAccount(accessToken);
+    this.save();
+    return this.digitalOceanAccount;
   }
 
+  /** See {@link CloudAccounts#connectGcpAccount} */
+  connectGcpAccount(refreshToken: string): gcp.Account {
+    this.gcpAccount = this.createGcpAccount(refreshToken);
+    this.save();
+    return this.gcpAccount;
+  }
+
+  /** See {@link CloudAccounts#disconnectDigitalOceanAccount} */
   disconnectDigitalOceanAccount(): void {
-    this.storage.removeItem(this.DIGITALOCEAN_TOKEN_STORAGE_KEY);
+    this.digitalOceanAccount = null;
+    this.save();
+  }
+
+  /** See {@link CloudAccounts#disconnectGcpAccount} */
+  disconnectGcpAccount(): void {
+    this.gcpAccount = null;
+    this.save();
   }
 
+  /** See {@link CloudAccounts#getDigitalOceanAccount} */
   getDigitalOceanAccount(): digitalocean.Account {
-    const accessToken = this.storage.getItem(this.DIGITALOCEAN_TOKEN_STORAGE_KEY);
-    return accessToken ? this.digitalOceanAccountFactory(accessToken) : null;
+    return this.digitalOceanAccount;
+  }
+
+  /** See {@link CloudAccounts#getGcpAccount} */
+  getGcpAccount(): gcp.Account {
+    return this.gcpAccount;
+  }
+
+  /** Loads the saved cloud accounts from disk. */
+  private load(): void {
+    const digitalOceanAccountJsonString =
+        this.storage.getItem(this.DIGITALOCEAN_ACCOUNT_STORAGE_KEY);
+    if (!digitalOceanAccountJsonString) {
+      const digitalOceanToken = this.loadLegacyDigitalOceanToken();
+      if (digitalOceanToken) {
+        this.digitalOceanAccount = this.createDigitalOceanAccount(digitalOceanToken);
+        this.save();
+      }
+    } else {
+      const digitalOceanAccountJson: DigitalOceanAccountJson =
+          JSON.parse(digitalOceanAccountJsonString);
+      this.digitalOceanAccount =
+          this.createDigitalOceanAccount(digitalOceanAccountJson.accessToken);
+    }
+
+    const gcpAccountJsonString = this.storage.getItem(this.GCP_ACCOUNT_STORAGE_KEY);
+    if (gcpAccountJsonString) {
+      const gcpAccountJson: GcpAccountJson =
+          JSON.parse(this.storage.getItem(this.GCP_ACCOUNT_STORAGE_KEY));
+      this.gcpAccount = this.createGcpAccount(gcpAccountJson.refreshToken);
+    }
+  }
+
+  /** Loads legacy DigitalOcean access token. */
+  private loadLegacyDigitalOceanToken(): string {
+    return this.storage.getItem(this.LEGACY_DIGITALOCEAN_STORAGE_KEY);
+  }
+
+  /** Replace the legacy DigitalOcean access token. */
+  private saveLegacyDigitalOceanToken(accessToken: string): void {
+    this.storage.setItem(this.LEGACY_DIGITALOCEAN_STORAGE_KEY, accessToken);
+  }
+
+  private createDigitalOceanAccount(accessToken: string): DigitalOceanAccount {
+    return new DigitalOceanAccount(accessToken, this.shadowboxSettings, this.isDebugMode);
+  }
+
+  private createGcpAccount(refreshToken: string): GcpAccount {
+    return new GcpAccount(refreshToken);
+  }
+
+  private save(): void {
```

Comment:
I think save is broken. You should delete the storage entries when the accounts are no longer present.

Reply:
Thanks. I addressed the issue.